File: src/print_partner/core/export_3mf.py
"""Export kit as 3MF with multi-printer plate layout and filament colors."""

from __future__ import annotations

import json
import re
import zipfile
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

import lib3mf
import trimesh

from print_partner.core.filament_assigner import PartCopy, assign_parts_to_printers
from print_partner.core.merge import MergePart
from print_partner.core.mesh_color import resolve_mesh_color
from print_partner.core.plate_packer import PlateLayout, PlacedItem, pack_copies_on_printer
from print_partner.core.printer_fleet import PrinterMachine

ProgressCallback = Callable[[int, int, str], None]

ExportLayoutMode = Literal["per_plate", "zip", "single_offset", "single_plate_only"]

_INVALID_XML_CHARS = re.compile(r"[^\w\s.\-()+]")
_PLATE_GAP_MM = 20.0


@dataclass
class Export3mfOptions:
    layout_mode: ExportLayoutMode = "per_plate"
    spacing_mm: float = 4.0
    enabled_printers: list[PrinterMachine] = field(default_factory=list)
    # When set, export uses these plate positions instead of auto-packing.
    plate_layouts: list[tuple[PrinterMachine, PlateLayout]] | None = None


@dataclass
class Export3mfResult:
    primary_path: Path
    paths: list[Path]
    object_count: int
    plate_count: int
    warnings: list[str]
    printer_summaries: list[str]


def sanitize_3mf_object_name(name: str) -> str:
    base = Path(name).name.strip() or "part.stl"
    cleaned = _INVALID_XML_CHARS.sub("_", base)
    return cleaned[:200] if cleaned else "part.stl"


def object_display_name(filename: str, unit: int, used_names: set[str]) -> str:
    base = sanitize_3mf_object_name(filename)
    if unit <= 1:
        display = base
    else:
        stem = Path(base).stem
        suffix = Path(base).suffix or ".stl"
        display = sanitize_3mf_object_name(f"{stem}{suffix} ({unit})")

    if display not in used_names:
        used_names.add(display)
        return display

    stem = Path(base).stem
    suffix = Path(base).suffix or ".stl"
    n = 2
    while True:
        candidate = sanitize_3mf_object_name(f"{stem}{suffix} ({n})")
        if candidate not in used_names:
            used_names.add(candidate)
            return candidate
        n += 1


def _filament_material_key(part: MergePart) -> tuple[str, str]:
    label = (part.filament_display or "").strip() or part.role
    fid = part.filament_color_id or label
    return fid, label


def _hex_to_lib3mf_color(wrapper: lib3mf.Wrapper, hex_color: str) -> object:
    h = hex_color if hex_color.startswith("#") else f"#{hex_color.lstrip('#')}"
    return wrapper.RGBAToColor(int(h[1:3], 16), int(h[3:5], 16), int(h[5:7], 16), 255)


def _lib3mf_position(x: float, y: float, z: float) -> lib3mf.Position:
    p = lib3mf.Position()
    p.Coordinates[0] = float(x)
    p.Coordinates[1] = float(y)
    p.Coordinates[2] = float(z)
    return p


def _lib3mf_triangle(a: int, b: int, c: int) -> lib3mf.Triangle:
    t = lib3mf.Triangle()
    t.Indices[0] = int(a)
    t.Indices[1] = int(b)
    t.Indices[2] = int(c)
    return t


def _add_trimesh_to_object(mesh_object: object, mesh: trimesh.Trimesh) -> None:
    for v in mesh.vertices:
        mesh_object.AddVertex(_lib3mf_position(v[0], v[1], v[2]))
    for f in mesh.faces:
        mesh_object.AddTriangle(_lib3mf_triangle(f[0], f[1], f[2]))


def _write_items_to_model(
    model: object,
    wrapper: lib3mf.Wrapper,
    items: list[PlacedItem],
    used_names: set[str],
    material_by_key: dict[str, tuple[object, int]],
    x_offset: float = 0.0,
) -> int:
    count = 0
    for placed in items:
        copy = placed.copy
        part = copy.part
        mesh = placed.mesh
        if x_offset:
            mesh = mesh.copy()
            mesh.apply_translation([x_offset, 0.0, 0.0])

        display_name = object_display_name(part.filename, copy.unit, used_names)
        mesh_object = model.AddMeshObject()
        mesh_object.SetName(display_name)
        mesh_object.SetPartNumber(display_name)
        _add_trimesh_to_object(mesh_object, mesh)

        key, label = _filament_material_key(part)
        if key not in material_by_key:
            group = model.AddBaseMaterialGroup()
            display_hex = resolve_mesh_color(part.role, part.filament_hex)
            prop_id = group.AddMaterial(label, _hex_to_lib3mf_color(wrapper, display_hex))
            material_by_key[key] = (group, prop_id)
        group, prop_id = material_by_key[key]
        mesh_object.SetObjectLevelProperty(group.GetResourceID(), prop_id)

        build_item = model.AddBuildItem(mesh_object, wrapper.GetIdentityTransform())
        build_item.SetPartNumber(display_name)
        count += 1
    return count


def _write_plate_file(
    path: Path,
    items: list[PlacedItem],
    x_offset: float = 0.0,
) -> int:
    wrapper = lib3mf.get_wrapper()
    model = wrapper.CreateModel()
    material_by_key: dict[str, tuple[object, int]] = {}
    used_names: set[str] = set()
    count = _write_items_to_model(model, wrapper, items, used_names, material_by_key, x_offset)
    if count > 0:
        model.QueryWriter("3mf").WriteToFile(str(path))
    return count


def _slug(name: str) -> str:
    return re.sub(r"[^\w\-.]+", "_", name.replace(" ", "_"))


def export_profile_3mf(
    profile_name: str,
    parts: list[MergePart],
    exports_dir: Path,
    on_progress: ProgressCallback | None = None,
    *,
    cancel_check: Callable[[], bool] | None = None,
    options: Export3mfOptions | None = None,
) -> Export3mfResult:
    """Export using print plan printers, plate packing, and layout mode."""
    opts = options or Export3mfOptions()
    safe_profile = _slug(profile_name)
    output_dir = exports_dir / safe_profile / "3mf_export"
    output_dir.mkdir(parents=True, exist_ok=True)

    included = [p for p in parts if p.included]
    missing_path = [p for p in included if not p.absolute_path or not p.absolute_path.is_file()]
    exportable = [p for p in included if p.absolute_path and p.absolute_path.is_file()]

    warnings: list[str] = [
        f"Missing STL: {p.relative_path} ({p.source_layer})" for p in missing_path
    ]

    copies: list[PartCopy] = []
    for part in exportable:
        qty = max(1, part.quantity_effective)
        for unit in range(1, qty + 1):
            copies.append(PartCopy(part=part, unit=unit))

    empty = Export3mfResult(
        primary_path=output_dir / f"{safe_profile}.3mf",
        paths=[],
        object_count=0,
        plate_count=0,
        warnings=warnings,
        printer_summaries=[],
    )
    if not copies:
        return empty

    printers = opts.enabled_printers
    if not printers:
        warnings.append("No printers enabled. Configure printers on the Print tab.")
        return empty

    if opts.plate_layouts:
        all_plates = list(opts.plate_layouts)
        total_work = sum(len(plate.items) for _, plate in all_plates)
        done = 0
        for _printer, plate in all_plates:
            for item in plate.items:
                done += 1
                if cancel_check and cancel_check():
                    break
                if on_progress:
                    on_progress(
                        done,
                        max(1, total_work),
                        item.copy.part.filename if plate.items else "",
                    )
    else:
        by_printer, assign_warnings = assign_parts_to_printers(copies, printers)
        warnings.extend(assign_warnings)

        all_plates = []
        total_work = sum(len(by_printer[p.id]) for p in printers)
        done = 0

        for printer in printers:
            pcopies = by_printer.get(printer.id, [])
            if not pcopies:
                continue
            plates, pack_warnings = pack_copies_on_printer(
                printer, pcopies, spacing_mm=opts.spacing_mm
            )
            warnings.extend(pack_warnings)
            for plate in plates:
                all_plates.append((printer, plate))
                for _ in plate.items:
                    done += 1
                    if cancel_check and cancel_check():
                        break
                    if on_progress:
                        on_progress(
                            done,
                            max(1, total_work),
                            plate.items[0].copy.part.filename if plate.items else "",
                        )

    if not all_plates:
        return empty

    if opts.layout_mode == "single_plate_only":
        multi = sum(1 for _, p in printers if len([pl for pr, pl in all_plates if pr.id == p.id]) > 1)
        if multi > 0 or len(all_plates) > 1:
            warnings.append(
                "Single-plate export requires everything to fit on one bed. "
                "Use per-plate or zip mode, or enable a larger printer."
            )
            return empty

    paths: list[Path] = []
    object_count = 0
    manifest_printers: list[dict] = []
    summaries: list[str] = []

    if opts.layout_mode == "single_offset":
        wrapper = lib3mf.get_wrapper()
        model = wrapper.CreateModel()
        material_by_key: dict[str, tuple[object, int]] = {}
        used_names: set[str] = set()
        x_cursor = 0.0
        for printer, plate in all_plates:
            object_count += _write_items_to_model(
                model, wrapper, plate.items, used_names, material_by_key, x_cursor
            )
            x_cursor += printer.bed_width_mm + _PLATE_GAP_MM
        out_path = output_dir / f"{safe_profile}.3mf"
        if object_count > 0:
            model.QueryWriter("3mf").WriteToFile(str(out_path))
            paths.append(out_path)
    else:
        for printer, plate in all_plates:
            slug_printer = _slug(printer.name)
            fname = f"{safe_profile}_{slug_printer}_plate_{plate.index:02d}.3mf"
            out_path = output_dir / fname
            n = _write_plate_file(out_path, plate.items)
            object_count += n
            if n > 0:
                paths.append(out_path)

    for printer in printers:
        printer_plates = [pl for pr, pl in all_plates if pr.id == printer.id]
        if not printer_plates:
            continue
        summaries.append(f"{printer.name}: {len(printer_plates)} plate(s)")
        plate_entries = []
        from print_partner.core.plate_preview import group_plate_items_by_source

        for plate in printer_plates:
            groups = group_plate_items_by_source(plate.items)
            part_names: list[str] = []
            for group in groups:
                part_names.extend(group.part_names)
            fname = f"{safe_profile}_{_slug(printer.name)}_plate_{plate.index:02d}.3mf"
            plate_entries.append(
                {
                    "file": fname,
                    "parts": part_names,
                    "groups": [
                        {
                            "repo": g.repo,
                            "folder": g.folder,
                            "parts": g.part_names,
                        }
                        for g in groups
                    ],
                }
            )
        manifest_printers.append(
            {
                "id": printer.id,
                "name": printer.name,
                "bed_mm": [printer.bed_width_mm, printer.bed_depth_mm],
                "plates": plate_entries,
            }
        )

    manifest = {
        "kit": profile_name,
        "layout_mode": opts.layout_mode,
        "printers": manifest_printers,
        "warnings": warnings,
    }
    manifest_path = output_dir / "print_plan.json"
    manifest_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")

    if opts.layout_mode == "zip" and paths:
        zip_path = output_dir / f"{safe_profile}_plates.zip"
        with zipfile.ZipFile(zip_path, "w", compression=zipfile.ZIP_DEFLATED) as zf:
            for p in paths:
                zf.write(p, arcname=p.name)
            zf.write(manifest_path, arcname="print_plan.json")

    primary = paths[0] if paths else output_dir / f"{safe_profile}.3mf"
    if opts.layout_mode == "zip":
        zip_path = output_dir / f"{safe_profile}_plates.zip"
        if zip_path.is_file():
            primary = zip_path

    return Export3mfResult(
        primary_path=primary,
        paths=paths,
        object_count=object_count,
        plate_count=len(all_plates),
        warnings=warnings,
        printer_summaries=summaries,
    )


# Backward-compatible tuple API for simple callers
def export_profile_3mf_legacy(
    profile_name: str,
    parts: list[MergePart],
    exports_dir: Path,
    on_progress: ProgressCallback | None = None,
    *,
    cancel_check: Callable[[], bool] | None = None,
) -> tuple[Path, int, list[str]]:
    from print_partner.core.printer_fleet import load_fleet

    fleet = load_fleet()
    if not fleet:
        from print_partner.core.printer_fleet import load_printer_presets, new_machine_from_preset

        preset = load_printer_presets()[-1]
        fleet = [new_machine_from_preset(preset, "Default")]
    opts = Export3mfOptions(enabled_printers=fleet[:1])
    result = export_profile_3mf(
        profile_name, parts, exports_dir, on_progress, cancel_check=cancel_check, options=opts
    )
    return result.primary_path, result.object_count, result.warnings

File: web/packages/domain/src/export-3mf.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename, join } from "node:path";
import { zipSync, strToU8 } from "fflate";
import type { PartCopy } from "./checkoff-missing.js";
import { quantityEffective } from "./merge.js";
import { assignPartsToPrinters, type MergePartExport, type PrinterMachine } from "./filament-assigner.js";
import { resolveMeshColor } from "./mesh-color.js";
import { packCopiesOnPrinter, type PlateLayout, type PlacedItem } from "./plate-packer.js";
import { resolveLayoutToPlates } from "./plate-plan.js";
import type { KitPlateLayout } from "./plate-plan.js";
import { translateMesh } from "./stl-mesh.js";
import { profileExportDir, safePlanSlug } from "./export-paths.js";

const INVALID_XML_CHARS = /[^\w\s.\-()+]/g;
const PLATE_GAP_MM = 20;

export type ExportLayoutMode = "per_plate" | "zip" | "single_offset" | "single_plate_only";

export type Export3mfOptions = {
  layout_mode?: ExportLayoutMode;
  spacing_mm?: number;
  enabled_printers?: PrinterMachine[];
  plate_layouts?: Array<[PrinterMachine, PlateLayout]> | null;
  missing_only?: boolean;
  completed_by_match_key?: Record<string, boolean[]> | null;
};

export type Export3mfResult = {
  primary_path: string;
  paths: string[];
  object_count: number;
  plate_count: number;
  warnings: string[];
  printer_summaries: string[];
};

export function sanitize3mfObjectName(name: string): string {
  const base = basename(name.trim() || "part.stl");
  const cleaned = base.replace(INVALID_XML_CHARS, "_");
  return cleaned.slice(0, 200) || "part.stl";
}

export function objectDisplayName(filename: string, unit: number, usedNames: Set<string>): string {
  const base = sanitize3mfObjectName(filename);
  let display = base;
  if (unit > 1) {
    const dot = base.lastIndexOf(".");
    const stem = dot >= 0 ? base.slice(0, dot) : base;
    const suffix = dot >= 0 ? base.slice(dot) : ".stl";
    display = sanitize3mfObjectName(`${stem}${suffix} (${unit})`);
  }
  if (!usedNames.has(display)) {
    usedNames.add(display);
    return display;
  }
  const dot = base.lastIndexOf(".");
  const stem = dot >= 0 ? base.slice(0, dot) : base;
  const suffix = dot >= 0 ? base.slice(dot) : ".stl";
  let n = 2;
  while (true) {
    const candidate = sanitize3mfObjectName(`${stem}${suffix} (${n})`);
    if (!usedNames.has(candidate)) {
      usedNames.add(candidate);
      return candidate;
    }
    n += 1;
  }
}

function filamentMaterialKey(part: MergePartExport): [string, string] {
  const label = (part.filamentDisplay ?? part.filament_display ?? "").trim() || part.role;
  const fid = part.filamentColorId ?? part.filament_color_id ?? label;
  return [fid, label];
}

function escapeXml(text: string): string {
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function hexToDisplayColor(hex: string): string {
  const h = hex.startsWith("#") ? hex.slice(1) : hex;
  if (h.length !== 6) return "#808080FF";
  return `#${h}FF`;
}

type MeshObjectSpec = {
  id: number;
  name: string;
  mesh: { vertices: Array<[number, number, number]>; faces: Array<[number, number, number]> };
  materialId: number;
  xOffset?: number;
};

function buildModelXml(objects: MeshObjectSpec[], materials: Map<string, { id: number; label: string; hex: string }>): string {
  const matEntries = [...materials.values()];
  let resources = "";
  for (const mat of matEntries) {
    resources += `    <basematerials id="${mat.id}">\n`;
    resources += `      <base name="${escapeXml(mat.label)}" displaycolor="${hexToDisplayColor(mat.hex)}"/>\n`;
    resources += `    </basematerials>\n`;
  }
  for (const obj of objects) {
    resources += `    <object id="${obj.id}" type="model">\n`;
    resources += `      <mesh>\n        <vertices>\n`;
    for (const [x, y, z] of obj.mesh.vertices) {
      resources += `          <vertex x="${x}" y="${y}" z="${z}"/>\n`;
    }
    resources += `        </vertices>\n        <triangles>\n`;
    for (const [v1, v2, v3] of obj.mesh.faces) {
      resources += `          <triangle v1="${v1}" v2="${v2}" v3="${v3}"/>\n`;
    }
    resources += `        </triangles>\n      </mesh>\n    </object>\n`;
  }

  let build = "";
  for (const obj of objects) {
    const pid = obj.materialId;
    build += `    <item objectid="${obj.id}"`;
    if (pid) build += ` pid="${pid}" pindex="1"`;
    build += `/>\n`;
  }

  return `<?xml version="1.0" encoding="UTF-8"?>
<model unit="millimeter" xml:lang="en-US" xmlns="http://schemas.microsoft.com/3dmanufacturing/core/2015/02" xmlns:m="http://schemas.microsoft.com/3dmanufacturing/material/2015/02">
  <resources>
${resources}  </resources>
  <build>
${build}  </build>
</model>`;
}

function write3mfZip(outPath: string, modelXml: string): void {
  const modelBytes = strToU8(modelXml);
  const contentTypes = `<?xml version="1.0" encoding="UTF-8"?>
<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>
  <Default Extension="model" ContentType="application/vnd.ms-package.3dmanufacturing-3dmodel+xml"/>
</Types>`;
  const rels = `<?xml version="1.0" encoding="UTF-8"?>
<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
  <Relationship Target="/3D/3dmodel.model" Id="rel0" Type="http://schemas.microsoft.com/3dmanufacturing/2013/01/3dmodel"/>
</Relationships>`;
  const plateMeta = JSON.stringify({
    plate_index: 1,
    object_ids: [],
    outside: false,
  });
  const sliceInfo = `<?xml version="1.0" encoding="UTF-8"?>
<config xmlns="http://schemas.bambulab.com/package/2021" version="1.0">
  <header>
    <printer>Bambu Lab X1</printer>
    <print_time>0</print_time>
  </header>
</config>`;
  const zipped = zipSync({
    "[Content_Types].xml": strToU8(contentTypes),
    "_rels/.rels": strToU8(rels),
    "3D/3dmodel.model": modelBytes,
    "Metadata/plate_1.json": strToU8(plateMeta),
    "Metadata/slice_info.config": strToU8(sliceInfo),
  });
  writeFileSync(outPath, zipped);
}

function itemsToObjects(
  items: PlacedItem[],
  usedNames: Set<string>,
  materialByKey: Map<string, { id: number; label: string; hex: string }>,
  xOffset = 0,
): MeshObjectSpec[] {
  const specs: MeshObjectSpec[] = [];
  let nextId = 2;
  let nextMatId = 10;
  for (const placed of items) {
    const part = placed.copy.part as MergePartExport;
    let mesh = placed.mesh;
    if (xOffset) mesh = translateMesh(mesh, xOffset, 0, 0);
    const displayName = objectDisplayName(part.filename, placed.copy.unit, usedNames);
    const [key, label] = filamentMaterialKey(part);
    let mat = materialByKey.get(key);
    if (!mat) {
      const hex = resolveMeshColor(part.role, part.filamentHex ?? part.filament_hex);
      mat = { id: nextMatId++, label, hex };
      materialByKey.set(key, mat);
    }
    specs.push({
      id: nextId++,
      name: displayName,
      mesh: { vertices: mesh.vertices, faces: mesh.faces },
      materialId: mat.id,
    });
  }
  return specs;
}

function writePlateFile(path: string, items: PlacedItem[], xOffset = 0): number {
  const usedNames = new Set<string>();
  const materialByKey = new Map<string, { id: number; label: string; hex: string }>();
  const objects = itemsToObjects(items, usedNames, materialByKey, xOffset);
  if (!objects.length) return 0;
  const xml = buildModelXml(objects, materialByKey);
  write3mfZip(path, xml);
  return objects.length;
}

export function exportProfile3mf(
  profileName: string,
  parts: MergePartExport[],
  exportsDir: string,
  options: Export3mfOptions = {},
): Export3mfResult {
  const opts = options;
  const layoutMode = opts.layout_mode ?? "per_plate";
  const spacingMm = opts.spacing_mm ?? 4;
  const safeProfile = safePlanSlug(profileName);
  const outputDir = profileExportDir(exportsDir, profileName, "3mf");
  mkdirSync(outputDir, { recursive: true });

  const included = parts.filter((p) => p.included);
  const missingPath = included.filter((p) => !p.absolutePath);
  const exportable = included.filter((p) => p.absolutePath);

  const warnings: string[] = missingPath.map(
    (p) => `Missing STL: ${p.relativePath} (${p.sourceLayer})`,
  );

  const copies: PartCopy[] = [];
  const completed = opts.completed_by_match_key ?? {};
  for (const part of exportable) {
    const qty = Math.max(1, part.quantityEffective ?? part.quantity_effective ?? quantityEffective(part));
    const units = completed[part.matchKey];
    for (let unit = 1; unit <= qty; unit++) {
      if (opts.missing_only) {
        const idx = unit - 1;
        if (units && idx < units.length && units[idx]) continue;
      }
      copies.push({ part, unit });
    }
  }
  if (opts.missing_only && !copies.length) {
    warnings.push("All included units are already marked printed in checkoff.");
  }

  const empty: Export3mfResult = {
    primary_path: join(outputDir, `${safeProfile}.3mf`),
    paths: [],
    object_count: 0,
    plate_count: 0,
    warnings,
    printer_summaries: [],
  };
  if (!copies.length) return empty;

  const printers = opts.enabled_printers ?? [];
  if (!printers.length) {
    warnings.push("No printers enabled. Configure printers on the Print tab.");
    return empty;
  }

  let allPlates: Array<[PrinterMachine, PlateLayout]> = [];
  if (opts.plate_layouts) {
    allPlates = opts.plate_layouts;
  } else {
    const [byPrinter, assignWarnings] = assignPartsToPrinters(copies, printers);
    warnings.push(...assignWarnings);
    for (const printer of printers) {
      const pcopies = byPrinter[printer.id] ?? [];
      if (!pcopies.length) continue;
      const [plates, packWarnings] = packCopiesOnPrinter(printer, pcopies, { spacing_mm: spacingMm });
      warnings.push(...packWarnings);
      for (const plate of plates) allPlates.push([printer, plate]);
    }
  }

  if (!allPlates.length) return empty;

  if (layoutMode === "single_plate_only") {
    const multi = printers.filter(
      (p) => allPlates.filter(([pr]) => pr.id === p.id).length > 1,
    ).length;
    if (multi > 0 || allPlates.length > 1) {
      warnings.push(
        "Single-plate export requires everything to fit on one bed. " +
          "Use per-plate or zip mode, or enable a larger printer.",
      );
      return empty;
    }
  }

  const paths: string[] = [];
  let objectCount = 0;

  if (layoutMode === "single_offset") {
    const usedNames = new Set<string>();
    const materialByKey = new Map<string, { id: number; label: string; hex: string }>();
    let xOffset = 0;
    const allObjects: MeshObjectSpec[] = [];
    for (const [printer, plate] of allPlates) {
      const objs = itemsToObjects(plate.items, usedNames, materialByKey, xOffset);
      allObjects.push(...objs);
      objectCount += objs.length;
      xOffset += printer.bed_width_mm + PLATE_GAP_MM;
    }
    const outPath = join(outputDir, `${safeProfile}.3mf`);
    if (allObjects.length) {
      write3mfZip(outPath, buildModelXml(allObjects, materialByKey));
      paths.push(outPath);
    }
  } else {
    for (const [printer, plate] of allPlates) {
      const slugPrinter = safePlanSlug(printer.name);
      const slugGroup = plate.group_label ? `_${safePlanSlug(plate.group_label).slice(0, 80)}` : "";
      const fname = plate.group_label
        ? `${safeProfile}_${slugPrinter}${slugGroup}_p${String(plate.index).padStart(2, "0")}.3mf`
        : `${safeProfile}_${slugPrinter}_plate_${String(plate.index).padStart(2, "0")}.3mf`;
      const outPath = join(outputDir, fname);
      const n = writePlateFile(outPath, plate.items);
      objectCount += n;
      if (n > 0) paths.push(outPath);
    }
  }

  const summaries: string[] = [];
  for (const printer of printers) {
    const printerPlates = allPlates.filter(([pr]) => pr.id === printer.id);
    if (!printerPlates.length) continue;
    summaries.push(`${printer.name}: ${printerPlates.length} plate(s)`);
  }

  const manifest = {
    kit: profileName,
    layout_mode: layoutMode,
    printers: summaries,
    warnings,
  };
  const manifestPath = join(outputDir, "print_plan.json");
  writeFileSync(manifestPath, JSON.stringify(manifest, null, 2), "utf8");

  let primary = paths[0] ?? join(outputDir, `${safeProfile}.3mf`);
  if (layoutMode === "zip" && paths.length) {
    const zipEntries: Record<string, Uint8Array> = {
      "print_plan.json": strToU8(readFileSync(manifestPath, "utf8")),
    };
    for (const p of paths) {
      zipEntries[basename(p)] = readFileSync(p);
    }
    const zipPath = join(outputDir, `${safeProfile}_plates.zip`);
    writeFileSync(zipPath, zipSync(zipEntries));
    primary = zipPath;
  }

  return {
    primary_path: primary,
    paths,
    object_count: objectCount,
    plate_count: allPlates.length,
    warnings,
    printer_summaries: summaries,
  };
}

export function exportProfile3mfWithLayout(
  profileName: string,
  parts: MergePartExport[],
  exportsDir: string,
  printers: PrinterMachine[],
  copies: PartCopy[],
  plateLayout: KitPlateLayout | null,
  options: Omit<Export3mfOptions, "enabled_printers"> = {},
): Export3mfResult {
  let plateLayouts: Array<[PrinterMachine, PlateLayout]> | undefined;
  if (plateLayout && copies.length) {
    const [resolved] = resolveLayoutToPlates(plateLayout, printers, copies);
    plateLayouts = resolved;
  }
  return exportProfile3mf(profileName, parts, exportsDir, {
    ...options,
    enabled_printers: printers,
    plate_layouts: plateLayouts ?? options.plate_layouts ?? undefined,
    spacing_mm: options.spacing_mm ?? plateLayout?.spacing_mm ?? 4,
  });
}
